### PR TITLE
Issue/2808 convert toes6 classes

### LINKED ISSUE #135: course.json correctToPass can be missing
Original issue: https://community.adaptlearning.org/mod/forum/discuss.php?d=4059#p14949

This seems to be as either the json hasn't been updated or the schema default hasn't been applied.

https://github.com/adaptlearning/adapt_framework/blob/b1b4dba7f53c18ed045a003997aed06dd5d9f66d/src/course/en/course.json#L131-L134 `_correctToPass` is missing in the default course.

If it is undefined it should reflect the `_scoreToPass`.

### PR DESCRIPTION
**Pull request by eleanor-heath**
part of https://github.com/adaptlearning/adapt_framework/issues/2808

### Changed
* converted to ees6 classes and modules

fixes #135

### Changed
* `_correctToPass` defaults to `_scoreToPass` if undefined on the major assessment

### Added
* console warn to say that the assessment is missing `_correctToPass` in the course.json

### CODE REVIEW

File: js/adapt-assessmentArticleModel.js [hunk 1]
```diff
@@ -24,809 +22,804 @@ define([
...
+  _onQuestionCompleted(questionModel, value) {
+    if (value === false) return;
+    if (!questionModel.get('_isInteractionComplete')) return;
+
+    let numberOfQuestionsAnswered = this.get('_numberOfQuestionsAnswered');
+    numberOfQuestionsAnswered++;
+    this.set('_numberOfQuestionsAnswered', numberOfQuestionsAnswered);
+
+    this._updateQuestionsState();
+    Assessment.saveState();
```
Comment (cahirodoherty-learningpool) [inline js/adapt-assessmentArticleModel.js:303]: Was this intentional? Are you sure it's still functioning?

Comment (eleanor-heath) [reply js/adapt-assessmentArticleModel.js]: Ive reverted that in the new commit

**Review by ethan-lp — APPROVED** [review]
👀

File: js/assessment.js [hunk 2]
```diff
@@ -1,409 +1,397 @@
...
-    getConfig: function () {
-      var assessmentsConfig = Adapt.course.get('_assessment');
+  saveState() {
 
-      if (assessmentsConfig && assessmentsConfig._isDefaultsLoaded) {
-        return assessmentsConfig;
-      }
+    this._saveStateModel = {};
+    for (let i = 0; i < this._assessments.length; i++) {
+      const assessmentModel = this._assessments[i];
```
Comment (tomgreenfield) [inline js/assessment.js:316]: `for...of` loops are nice in these situations:
```suggestion
    for (const assessmentModel of this._assessments) {
```

**Review by oliverfoster — APPROVED** [review]

File: js/adapt-assessmentArticleModel.js [hunk 3]
```diff
@@ -1,836 +1,826 @@
...
 
-      // stop resetting if not complete or not allowed
-      if (this.get('_assessmentCompleteInSession') &&
-        !assessmentConfig._isResetOnRevisit &&
-        !isPageReload &&
-        !force) {
+    if (!isPageReload) {
+      // only perform this section when not attempting to reload the page
+      // wait for reset to trigger
+      this.once('reset', () => {
```
Comment (tomgreenfield) [inline js/adapt-assessmentArticleModel.js:648]: Was it on purpose that an arrow function was introduced here but not on the subsequent lines, i.e. 657/661?

File: js/adapt-assessmentArticleModel.js [hunk 4]
```diff
@@ -1,836 +1,826 @@
...
-      }
+    return true;
+  },
+
+  getSaveState() {
+    const state = this.getState();
+    let blocks;
+    const cfg = this.getConfig();
+    const banksActive = cfg._banks && cfg._banks._isEnabled && cfg._banks._split.length > 1;
+    const randomisationActive = cfg._randomisation && cfg._randomisation._isEnabled;
```
Comment (tomgreenfield) [inline js/adapt-assessmentArticleModel.js:677]:
```suggestion
    const banksActive = cfg._banks?._isEnabled && cfg._banks._split.length > 1;
    const randomisationActive = cfg._randomisation?._isEnabled;
```

File: js/adapt-assessmentArticleView.js [hunk 5]
```diff
@@ -1,57 +1,53 @@
...
 
-    _setupEventListeners: function() {
-      this.listenTo(Adapt, {
-        'assessments:complete': this._onAssessmentComplete,
-        'assessments:reset': this._onAssessmentReset,
-        remove: this._onRemove
-      });
-    },
+      const config = this.model.getConfig();
+      if (config && config._questions && config._questions._canShowMarking === false) {
```
Comment (tomgreenfield) [inline js/adapt-assessmentArticleView.js:12]:
```suggestion
      if (config?._questions?._canShowMarking === false) {
```

File: js/adapt-assessmentQuestionBank.js [hunk 6]
```diff
@@ -1,71 +1,64 @@
...
+    this.checkResetUnunsedBlocks();
+
+    const questionBlocks = [];
+    const usedQuestionBlocks = this.usedQuestionBlocks.slice(0);
+
+    for (let i = 0; i < this._numQuestionBlocks; i++) {
+      let question = this.getRandomQuestion();
+      if (question !== undefined) {
+        questionBlocks.push(question);
+      } else {
```
Comment (tomgreenfield) [inline js/adapt-assessmentQuestionBank.js:31]: Could return early here with a `continue;`.

File: js/assessment.js [hunk 7]
```diff
@@ -1,409 +1,394 @@
...
+/*
+  * Here we setup a registry for all assessments
+  */
+const assessmentsConfigDefaults = {
+  _isPercentageBased: true,
+  _scoreToPass: 100,
+  _isDefaultsLoaded: true
+};
+
+export default Adapt.assessment = _.extend({
```
Comment (tomgreenfield) [inline js/assessment.js:12]: Suggest this is made into a controller, e.g.
```js
class Assessment extends Backbone.Controller {
  // ...
}

export default (Adapt.assessment = new Assessment());

File: js/assessment.js [hunk 8]
```diff
@@ -1,409 +1,394 @@
...
+    return states;
+  },
 
-        var assessments = this._assessments._byPageId[k];
+  _setPageProgress() {
+    // set _subProgressTotal and _subProgressComplete on pages that have assessment progress indicator requirements
 
-        var assessmentsTotal = assessments.length;
-        var assessmentsPassed = 0;
+    for (const k in this._assessments._byPageId) {
```
Comment (tomgreenfield) [inline js/assessment.js:200]:
```suggestion
    for (const [ id, assessments ] of Object.entries(this._assessments._byPageId)) {
```

File: js/assessment.js [hunk 9]
```diff
@@ -1,409 +1,394 @@
...
-      if (id === undefined) {
-        Adapt.log.warn('An assessment has been registered with an undefined value for "_id"');
-        return;
-      }
-
-      if (id === '') {
-        Adapt.log.warn('An assessment has been registered with an empty value for "_id"');
-      }
+      try {
+        const pageModel = Adapt.findById(k);
```
Comment (tomgreenfield) [inline js/assessment.js:218]:
```suggestion
        const pageModel = Adapt.findById(id);
```

File: js/assessment.js [hunk 10]
```diff
@@ -1,409 +1,394 @@
...
+        pageModel.set({
```
Comment (tomgreenfield) [inline js/assessment.js:219]: No longer require _try...catch_.
```suggestion
        pageModel?.set({
```

File: js/assessment.js [hunk 11]
```diff
@@ -1,409 +1,394 @@
...
+
+  _setupQuestionNumbering() {
+    const getRelatedQuestions = data => {
+      const currentAssessmentId = data._assessmentId;
+      const currentAssessment = Adapt.assessment.get(currentAssessmentId);
+      return currentAssessment.getState().questions;
+    };
+
+    Handlebars.registerHelper('questionNumber', function getQuestionNumber() {
+      const data = this.view ? this.view.model.toJSON() : this;
```
Comment (tomgreenfield) [inline js/assessment.js:255]:
```suggestion
      const data = this.view?.model.toJSON() || this;
```
See also line 264.

File: js/assessment.js [hunk 12]
```diff
@@ -1,409 +1,394 @@
...
+      Adapt.trigger('assessment:restored', this.getState());
+    }
+  },
 
-      Adapt.offlineStorage.set('a', this._saveStateModel);
-    },
+  get(id) {
+    if (id === undefined) {
+      return this._assessments.slice(0);
+    } else {
```
Comment (tomgreenfield) [inline js/assessment.js:305]: Unnecessary _else_.

File: js/assessment.js [hunk 13]
```diff
@@ -1,409 +1,394 @@
...
-        assessmentsConfig = $.extend(true, {}, assessmentsConfigDefaults, assessmentsConfig);
-      }
+    Adapt.offlineStorage.set('a', this._saveStateModel);
+  },
 
-      Adapt.course.set('_assessment', assessmentsConfig);
+  getConfig() {
+    let assessmentsConfig = Adapt.course.get('_assessment');
 
+    if (assessmentsConfig && assessmentsConfig._isDefaultsLoaded) {
```
Comment (tomgreenfield) [inline js/assessment.js:324]:
```suggestion
    if (assessmentsConfig?._isDefaultsLoaded) {
```

File: js/adapt-assessmentArticleExtension.js [hunk 14]
```diff
@@ -1,46 +1,41 @@
...
-  'core/js/models/articleModel',
-  './adapt-assessmentArticleView',
-  './adapt-assessmentArticleModel',
-  './assessment'
-], function(ArticleView, ArticleModel, AdaptAssessmentArticleView, AdaptAssessmentArticleModel) {
+import ArticleView from 'core/js/views/articleView';
+import ArticleModel from 'core/js/models/articleModel';
+import AdaptAssessmentArticleView from './adapt-assessmentArticleView';
+import AdaptAssessmentArticleModel from './adapt-assessmentArticleModel';
+
```
Comment (tomgreenfield) [inline js/adapt-assessmentArticleExtension.js:5]: Assessment code needs to be included.

```suggestion
import './assessment';

```

**Review by tomgreenfield — CHANGES_REQUESTED** [review]

File: js/adapt-assessmentArticleExtension.js (hunk 14 above)
Comment (oliverfoster) [reply js/adapt-assessmentArticleExtension.js:5]: sorry, this was my fault

File: js/assessment.js [hunk 15]
```diff
@@ -1,409 +1,389 @@
...
-    _isPercentageBased: true,
-    _scoreToPass: 100,
-    _isDefaultsLoaded: true
-  };
-
-  Adapt.assessment = _.extend({
-
-    // Private functions
-    _assessments: _.extend([], {
+import Backbone from 'backbone';
```
Comment (tomgreenfield) [inline js/assessment.js:1]:
```suggestion
```

**Review by tomgreenfield — APPROVED** [review]

**Pull request merged**